From 21a18c3fc7b2c471d548cfd5e6ade9743bfc8b2a Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Fri, 25 Apr 2014 15:27:03 -0400
Subject: [ata] ahci: fix Null pointer dereference in achi_host_active()

Message-id: <1398439625-8238-3-git-send-email-dmilburn@redhat.com>
Patchwork-id: 79181
O-Subject: [RHEL7.1 PATCH 2/4] ahci: fix Null pointer dereference in achi_host_active()
Bugzilla: 1067838
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

commit c91bc6ccd13254826fdfceddba0f3b5e308aa93e
Author: Xiaotian Feng <xtfeng@gmail.com>
Date:   Tue Jul 23 11:54:10 2013 +0800

    ahci: fix Null pointer dereference in achi_host_active()

    commit b29900e6 (AHCI: Make distinct names for ports in /proc/interrupts)
    introuded a regression, which resulted Null pointer dereference for achi
    host with dummy ports. For ahci ports, when the port is dummy port, its
    private_data will be NULL, as ata_dummy_port_ops doesn't support ->port_start.

    changes in v2: use pp to check dummy ports, update comments

    Reported-and-tested-by: Alex Williamson <alex.williamson@redhat.com>
    Signed-off-by: Xiaotian Feng <xtfeng@gmail.com>
    Signed-off-by: Tejun Heo <tj@kernel.org>
    Cc: Alexander Gordeev <agordeev@redhat.com>
    Cc: linux-ide@vger.kernel.org
    Cc: linux-kernel@vger.kernel.org

https://bugzilla.redhat.com/show_bug.cgi?id=1067838
https://brewweb.devel.redhat.com/taskinfo?taskID=7368411
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/ata/ahci.c b/drivers/ata/ahci.c
index bfa8db6..b6d33e1 100644
--- a/drivers/ata/ahci.c
+++ b/drivers/ata/ahci.c
@@ -1152,11 +1152,18 @@ int ahci_host_activate(struct ata_host *host, int irq, unsigned int n_msis)
   return rc;
 
  for (i = 0; i < host->n_ports; i++) {
+  const char* desc;
   struct ahci_port_priv *pp = host->ports[i]->private_data;
 
+  /* pp is NULL for dummy ports */
+  if (pp)
+   desc = pp->irq_desc;
+  else
+   desc = dev_driver_string(host->dev);
+
   rc = devm_request_threaded_irq(host->dev,
    irq + i, ahci_hw_interrupt, ahci_thread_fn, IRQF_SHARED,
-   pp->irq_desc, host->ports[i]);
+   desc, host->ports[i]);
   if (rc)
    goto out_free_irqs;
  }
-- 
1.7.1